Fix availability for manual trigger entities

gjohansson-ST opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix `ManualTriggerEntity` to not write state before calculating availability.

- Test availability first
- If available, render templates as normal
- If not available, no further rendering (all other properties remains as before)

Fair share of tests added.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/127960 fixes https://github.com/home-assistant/core/issues/99942 fixes https://github.com/home-assistant/core/issues/119992 fixes #133263
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/36910

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

gjohansson-ST committed:
Fix availability for manual trigger entities

gjohansson-ST commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -231,10 +232,26 @@ def _process_manual_data(self, value: Any | None = None) -> None:
+        if extra_state_attributes := self.extra_state_attributes:
+            attr.update(extra_state_attributes)
+
+        state = State(
```
We need to construct out own state to provide an uptodate `this` variable in the template processing

balloob replied:
Oof. Can't we just not support `this` for trigger based template entities. It's pretty ugly to make a fake state.

gjohansson-ST committed:
Fix when state breaks to stringify

flo-wer commented:
Does this also fix #99942 and #119992? Also does this replace #100079? If yes, it might be a good idea to copy the integration test for the rest sensor from your old PR.

gjohansson-ST committed:
Add test

gjohansson-ST commented:
> Does this also fix #99942 and #119992? Also does this replace #100079? If yes, it might be a good idea to copy the integration test for the rest sensor from your old PR.

Thanks @flo-wer I added those to be fixed by this PR and while this isn't specifically only for `rest` I guess including that test makes sense so I added it in.

emontnemery reviewed (changes requested):
For non-trigger template entities, we added the concept of "super template" which is always the availability template and which:
- Is rendered first
- Suppresses rendering of other templates if it's false

Can we do something similar here?

Also, we should preferably add direct tests of `ManualTriggerEntity`.
The existing test is in `tests/components/template/test_manual_trigger_entity.py`, it should be moved to `tests/helpers/test_trigger_template_entity.py`

emontnemery committed:
Try a different approach

MartinHjelmare commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -176,18 +176,43 @@ def restore_attributes(self, last_state: State) -> None:
 
+    def _render_availability_template(self, variables: dict[str, Any]) -> None:
+        """Render availability template."""
+        rendered = dict(self._static_rendered)
+        self._rendered = self._static_rendered
```
```suggestion
        static_rendered = self._static_rendered
        rendered = dict(static_rendered)
        self._rendered = static_rendered
```

MartinHjelmare replied:
Why do we set `self._rendered` here when we do it at the end of the method too?

MartinHjelmare commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -176,18 +176,43 @@ def restore_attributes(self, last_state: State) -> None:
+        """Render availability template."""
+        rendered = dict(self._static_rendered)
+        self._rendered = self._static_rendered
+        try:
+            key = CONF_AVAILABILITY
```
```suggestion
        key = CONF_AVAILABILITY
        try:
```

emontnemery commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -231,16 +255,22 @@ def _process_manual_data(self, value: Any | None = None) -> None:
-        self.async_write_ha_state()
+        run_variables: dict[str, Any] = {"value": value}
         this = None
         if state := self.hass.states.get(self.entity_id):
             this = state.as_dict()
-
-        run_variables: dict[str, Any] = {"value": value}
         # Silently try if variable is a json and store result in `value_json` if it is.
         with contextlib.suppress(*JSON_DECODE_EXCEPTIONS):
             run_variables["value_json"] = json_loads(run_variables["value"])
         variables = {"this": this, **(run_variables or {})}
+        self._render_availability_template(variables)
+
+        self.async_write_ha_state()
+        this = None
+        if state := self.hass.states.get(self.entity_id):
+            this = state.as_dict()
+
+        variables["this"] = this
```
I think the existing code is strange, why do we unconditionally call `self.async_write_ha_state` before rendering any templates. I'm guessing it has to do with startup behavior, before the entity is first added. However, the solution to that is to use `TemplateStateFromEntityId`, as is done in `TemplateEntity`: https://github.com/home-assistant/core/blob/d9eada044ccf28b6d8ebea0db637cae3b827ccc7/homeassistant/components/template/template_entity.py#L350

I would suggest we clean that up in an initial PR, then rebase this PR after.

emontnemery replied:
I also think the similar existing code in the template integration's `TriggerEntity` should be changed to use `TemplateStateFromEntityId` here: https://github.com/home-assistant/core/blob/d9eada044ccf28b6d8ebea0db637cae3b827ccc7/homeassistant/components/template/trigger_entity.py#L44-L48

gjohansson-ST replied:
> I think the existing code is strange, why do we unconditionally call self.async_write_ha_state before rendering any templates.

Because otherwise `this` would refer to the previous state instead of the current state.
With the current suggestion it's the same thing, but now it's moved to https://github.com/home-assistant/core/pull/128054/files#diff-414fea43d4fb87d5367be0a224750dbd804079ab4da80743243af0c66808e962R268 instead (so it considers the availability prior which is the issue to solve)

emontnemery replied:
> Because otherwise this would refer to the previous state instead of the current state.

That's how it's supposed to work. The `this` object is supposed to make it easy to access the current state when calculating a new state: https://github.com/home-assistant/core/pull/65201

emontnemery replied:
Related PR to handle edge case for new entity: https://github.com/home-assistant/core/pull/70911

gjohansson-ST commented:
Created #130134 #130135 #130136 as preliminary PR's for this one

gjohansson-ST committed:
Fix

gjohansson-ST committed:
Mod

gjohansson-ST committed:
Last fix

gjohansson-ST commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -176,18 +176,42 @@ def restore_attributes(self, last_state: State) -> None:
                 extra_state_attributes[attr] = last_state.attributes[attr]
             self._rendered[CONF_ATTRIBUTES] = extra_state_attributes
 
+    def _render_availability_template(self, variables: dict[str, Any]) -> None:
```
Should we change this to return a boolean to skip the rest of rendering in case it's not available?

gjohansson-ST committed:
Only render templates if availability is true

gjohansson-ST committed:
Fix

gjohansson-ST committed:
Fix

gjohansson-ST committed:
Remove render complex for availability

gjohansson-ST committed:
Mods

emontnemery commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -176,18 +177,41 @@ def restore_attributes(self, last_state: State) -> None:
 
             for key in self._to_render_complex:
+                if key == CONF_AVAILABILITY:
+                    continue
```
Can we add a test which exercises this path?

gjohansson-ST replied:
As with below comment, `availability` can not be complex, so removing this check.

emontnemery commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -121,6 +121,7 @@ def __init__(
         self._rendered = dict(self._static_rendered)
         self._parse_result = {CONF_AVAILABILITY}
         self._attr_device_class = config.get(CONF_DEVICE_CLASS)
+        self.error = False
```
Should this really be public? Also, can we give it a more descriptive name?

gjohansson-ST replied:
No, I renamed it

emontnemery commented on homeassistant/components/template/trigger_entity.py:
```diff
@@ -48,6 +48,7 @@ def _process_data(self) -> None:
             **(run_variables or {}),
         }
 
+        self._render_availability_template(variables)
         self._render_templates(variables)
```
Why do we need to call both methods? The `_render_templates` method anyway checks the state of availability template rendering, I think that method can also render the availability template?

gjohansson-ST replied:
I think that makes sense to have `render_templates` to directly call the `render_availability_template`

emontnemery commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -176,18 +177,41 @@ def restore_attributes(self, last_state: State) -> None:
+        rendered = {**self._static_rendered, **self._rendered}
+        key = CONF_AVAILABILITY
+        try:
+            if key in self._to_render_simple:
```
Don't we need to support complex availability templates?

gjohansson-ST replied:
According to my understanding from `__init__` it would appear that `availability`, ìcon`, `name` and `picture` can only be static or simple.

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/trigger_template_entity.py#L101-L112

emontnemery reviewed (commented):
Looks pretty good, some questions though.

Also, we should make sure the way the availability template suppresses rendering of other templates is documented, both for the trigger case and for the state case.

gjohansson-ST committed:
render templates calls availability render

gjohansson-ST committed:
Error better name

gjohansson-ST committed:
Remove availability from complex

gjohansson-ST committed:
Add state template test

gjohansson-ST committed:
Clean test

gjohansson-ST commented on homeassistant/helpers/trigger_template_entity.py:
```diff
@@ -176,12 +177,34 @@ def restore_attributes(self, last_state: State) -> None:
+                    variables,
+                    parse_result=key in self._parse_result,
+                )
+        except TemplateError as err:
+            logging.getLogger(f"{__package__}.{self.entity_id.split('.')[0]}").error(
+                "Error rendering %s template for %s: %s", key, self.entity_id, err
+            )
+            self._render_error = True
```
Here is a significant difference between state and trigger based templates.
If a state availability template fails to render, it assumes it's `available`
If a trigger availability template fails to render, it will go `unavailable`.

This PR does not change this behavior from before, but it's a notable difference on behavior which should be addressed.

frenck committed:
Merge branch 'dev' into manual_trigger_entity-fix-availability

emontnemery commented:
I think this can be closed due to the changes in #140660
